SU2_CFD/src/solvers/CNEMONSSolver.cpp

```diff
@@ -804,99 +792,137 @@ void CNEMONSSolver::BC_IsothermalCatalytic_Wall(CGeometry *geometry,
       Di   = nodes->GetDiffusionCoeff(iPoint);
       eves = nodes->GetEve(iPoint);
       hs   = FluidModel->ComputeSpeciesEnthalpy(Vi[T_INDEX], Vi[TVE_INDEX], eves);
-      for (iSpecies = 0; iSpecies < nSpecies; iSpecies++)
-        Yj[iSpecies] = Vj[RHOS_INDEX+iSpecies]/Vj[RHO_INDEX];
       rho    = Vi[RHO_INDEX];
       dTdU   = nodes->GetdTdU(iPoint);
       dTvedU = nodes->GetdTvedU(iPoint);
 
-      /*--- Calculate normal derivative of mass fraction ---*/
-      for (iSpecies = 0; iSpecies < nSpecies; iSpecies++)
-        dYdn[iSpecies] = (Yst[iSpecies]-Yj[iSpecies])/dij;
+      if (config->GetSupercatalytic_Wall()) {
 
-      /*--- Calculate supplementary quantities ---*/
-      SdYdn = 0.0;
-      for (iSpecies = 0; iSpecies < nSpecies; iSpecies++)
-        SdYdn += rho*Di[iSpecies]*dYdn[iSpecies];
+        const auto& Yst = config->GetSupercatalytic_Wall_Composition();
 
-      for (iSpecies = 0; iSpecies < nSpecies; iSpecies++) {
-        Res_Visc[iSpecies]         = -(-rho*Di[iSpecies]*dYdn[iSpecies]
-                                       +Yst[iSpecies]*SdYdn            )*Area;
-        Res_Visc[nSpecies+nDim]   += (Res_Visc[iSpecies]*hs[iSpecies]  )*Area;
-        Res_Visc[nSpecies+nDim+1] += (Res_Visc[iSpecies]*eves[iSpecies])*Area;
-      }
-
-      /*--- Viscous contribution to the residual at the wall ---*/
-      LinSysRes.SubtractBlock(iPoint, Res_Visc);
+        /*--- Calculate supplementary quantities ---*/
+        su2double dYdn;
+        su2double SdYdn = 0.0;
+        for (iSpecies = 0; iSpecies < nSpecies; iSpecies++) {
+          dYdn = (Yst[iSpecies]-Vj[RHOS_INDEX+iSpecies]/Vj[RHO_INDEX])/dij;
```

Comment:
maybe make dYdn a vector?

Reply:
Previously it was allocated as an array. I actually switched to this to eliminate the need to do a heap allocation each time this BC gets called, which should improve performance.